### PR TITLE
Allow query for efficiency scaling factor in off-process wells where required

### PR DESCRIPTION
**Pull request by vkip**
Fixes an issue with MPI hang when well_state[name] throws for off-process wells.

### CODE REVIEW

**Comment by vkip** [comment]
jenkins build this please

**Comment by vkip** [comment]
Failing tests are due to minor numerical differences and disappear, locally here at least, if setting the relaxed cnv tolerance equal to the strict cnv tolerance (--tolerance-cnv-relaxed=0.01).

**Comment by bska** [comment]
@akva2 : Would you mind having a look at this, please?

File: opm/simulators/wells/WellState.hpp [hunk 1]
```diff
@@ -180,6 +180,11 @@ class WellState
         return this->global_well_info.value().is_open(name);
     }
 
+    bool getGlobalEfficiencyScalingFactor(const std::string& name) const
```
Comment (akva2) [inline opm/simulators/wells/WellState.hpp:183]: took me a while to track this one. not a bool!

File: opm/simulators/wells/BlackoilWellModelGeneric.cpp [hunk 2]
```diff
@@ -1576,7 +1576,7 @@ calculateEfficiencyFactors(const int reportStepIdx)
     for (auto& well : well_container_generic_) {
         const Well& wellEcl = well->wellEcl();
         Scalar well_efficiency_factor = wellEcl.getEfficiencyFactor() *
-                                        wellState()[well->name()].efficiency_scaling_factor;
+                                        wellState().getGlobalEfficiencyScalingFactor(well->name());
```
Comment (blattms) [inline opm/simulators/wells/BlackoilWellModelGeneric.cpp:1579]: At other places there seems to be a check that the well is owned around this instead
```
if (this->well_state_.wellIsOwned(well.indexOfWell(), well.name())) {
  ...
```
Wouldn't this work here and the other places, too?

File: opm/simulators/wells/GasLiftStage2.cpp [hunk 3]
```diff
@@ -832,7 +832,7 @@ computeDelta(const std::string& well_name, bool add)
         if (this->well_state_.wellIsOwned(well.indexOfWell(), well_name)) {
             const auto& well_ecl = well.wellEcl();
             Scalar factor = well_ecl.getEfficiencyFactor() *
-                            this->well_state_[well_name].efficiency_scaling_factor;
+                            this->well_state_.getGlobalEfficiencyScalingFactor(well_name);
```
Comment (blattms) [inline opm/simulators/wells/GasLiftStage2.cpp:835]: Are you sure there was a problem here? 
Line 832 should prevent processing for wells that are not owned by us via `if (this->well_state_.wellIsOwned(well.indexOfWell(), well_name))`

Maybe we could use this at the other places, too?

**Review by blattms — COMMENTED** [review]
Without having looked much into this, I hope we can skip adding the global array.

At least for distributed wells the approach might count factors multiple times as is.

File: opm/simulators/wells/BlackoilWellModelGeneric.cpp (hunk 2 above)
Comment (blattms) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:1579]: Nope, it won't. (I was tricked by accumlate... below).
For distributed wells, we need to do this on all processes that have cells of a well.

File: opm/simulators/wells/WellState.hpp (hunk 1 above)
Comment (vkip) [reply opm/simulators/wells/WellState.hpp:183]: Wow, the joys of copy-paste to save a few keystrokes.. Thanks!

File: opm/simulators/wells/GasLiftStage2.cpp (hunk 3 above)
Comment (vkip) [reply opm/simulators/wells/GasLiftStage2.cpp:835]: You're right - got a bit too eager here.. will revert.

**Comment by akva2** [comment]
@vkip please have a look at https://github.com/OPM/opm-simulators/pull/5814

Maybe whatever you are trying to fix was caused by that? All the cases pass parallel regression tests (I only added one so we don't inflate the test suite too much).

File: opm/simulators/wells/GlobalWellInfo.hpp [hunk 4]
```diff
@@ -64,6 +64,7 @@ class GlobalWellInfo {
         comm.sum( this->m_in_injecting_group.data(), size);
         comm.sum( this->m_in_producing_group.data(), size);
         comm.sum( this->m_is_open.data(), size);
+        comm.min( this->m_efficiency_scaling_factors.data(), size);
```
[the remark this answers is not in the document]
Comment (vkip) [reply opm/simulators/wells/GlobalWellInfo.hpp:67]: Factors as initialized to 1, but may be set to something lower than 1 on the process that owns the well. Min to the this value on all processes (required a few places, e.g. in the network pressure calculations).

**Comment by vkip** [comment]
> @vkip please have a look at #5814
> 
> Maybe whatever you are trying to fix was caused by that? All the cases pass parallel regression tests (I only added one so we don't inflate the test suite too much).

Nope, both network and gas lift calculations need to get the efficiency factor for off process wells, in which case well_state[well_name] will throw..

**Comment by vkip** [comment]
jenkins build this please

**Comment by akva2** [comment]
okay, we definitely need both PRs then.

File: opm/simulators/wells/BlackoilWellModel_impl.hpp [hunk 5]
```diff
@@ -1061,7 +1061,8 @@ namespace Opm {
                                                                           this->well_open_times_,
                                                                           schedule_open))
                 {
-                    this->wellState()[wname].efficiency_scaling_factor = wscale;
+                    if (this->wellState().has(wname))
```
Comment (akva2) [inline opm/simulators/wells/BlackoilWellModel_impl.hpp:1064]: we shouldn't be here unless we have the well so this if should be redundant?

Comment (vkip) [reply opm/simulators/wells/BlackoilWellModel_impl.hpp:1064]: ah, right - will revert

File: opm/simulators/wells/BlackoilWellModelGeneric.cpp (hunk 2 above)
Comment (blattms) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:1579]: @akva2 I think this is where it broke and what @vkip is fixing.

Comment (vkip) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:1579]: Not sure I follow - as commented below, both network and gas lift calculations need to get the efficiency factor for off process wells.

Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:1579]: I don't think this block is actually where the real fix is situated. This loops over the well_container_generic_ and that only has entries for local active wells. The real reason for the change are the two call sites in GasLiftGroupInfo.

Comment (vkip) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:1579]: and this one: https://github.com/OPM/opm-simulators/blob/ee972dcbe2187c1051773415dda47f37f072f2b0/opm/simulators/wells/WellGroupHelpers.cpp#L875-L876

Comment (akva2) [reply opm/simulators/wells/BlackoilWellModelGeneric.cpp:1579]: indeed.

**Comment by vkip** [comment]
jenkins build this please

**Comment by akva2** [comment]
@vkip please pick and amend with https://github.com/akva2/opm-simulators/commit/c717f94a7a107eb136b001eb454221ac9007b8c1

**Comment by vkip** [comment]
> @vkip please pick and amend with [akva2@c717f94](https://github.com/akva2/opm-simulators/commit/c717f94a7a107eb136b001eb454221ac9007b8c1)

Pretty sure I'll mess that up - probably safer if you just create a new PR, so closing this one.

**Pull request closed without merging**